pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TopicsBase.java

```diff
@@ -0,0 +1,715 @@
+                TopicName topicName = TopicName.get(topic);
+                pulsar().getOwningTopics().get(topicName.getPartitionedTopicName())
+                        .remove(topicName.getPartitionIndex());
+            } else {
+                try {
+                    t.get().publishMessage(messageToByteBuf(message),
+                            RestMessagePublishContext.get(publishResult, t.get(), System.nanoTime()));
+                } catch (Exception e) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Fail to publish single messages to topic  {}: {} ",
+                                topicName, e.getCause());
+                    }
+                    publishResult.completeExceptionally(e);
+                }
+            }
+        });
+
+        return publishResult;
+    }
+
+    // Process results for all message publishing attempts
+    private void processPublishMessageResults(List<ProducerAck> produceMessageResults,
+                                              List<CompletableFuture<PositionImpl>> publishResults) {
+        // process publish message result
+        for (int index = 0; index < publishResults.size(); index++) {
+            try {
+                PositionImpl position = publishResults.get(index).get();
+                MessageId messageId = new MessageIdImpl(position.getLedgerId(), position.getEntryId(),
+                        Integer.parseInt(produceMessageResults.get(index).getMessageId()));
+                produceMessageResults.get(index).setMessageId(messageId.toString());
+            } catch (Exception e) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Fail publish [{}] message with rest produce message request for topic  {}: {} ",
+                            index, topicName);
+                }
+                if (e instanceof BrokerServiceException.TopicNotFoundException) {
+                    // Topic ownership might changed, force to look up again.
+                    pulsar().getOwningTopics().remove(topicName.getPartitionedTopicName());
+                }
+                extractException(e, produceMessageResults.get(index));
+            }
+        }
+    }
+
+    // Return error code depends on exception we got indicating if client should retry with same broker.
+    private void extractException(Exception e, ProducerAck produceMessageResult) {
+        if (!(e instanceof BrokerServiceException.TopicFencedException && e instanceof ManagedLedgerException)) {
+            produceMessageResult.setErrorCode(2);
+        } else {
+            produceMessageResult.setErrorCode(1);
+        }
+        produceMessageResult.setErrorMsg(e.getMessage());
+    }
+
+    // Look up topic owner for given topic. Return if asyncResponse has been completed
+    // which indicating redirect or exception.
+    private boolean findOwnerBrokerForTopic(boolean authoritative, AsyncResponse asyncResponse) {
```

Comment:
Why not use TopicLookupBase directly

Reply:
there're some custom logic to update the serving topics and partitions when completing the lookup, so can't directly use TopicLookupBase here.